[Linus Lee] iP

zupey opened the pull request:
# Duke v0.1
Use Duke to keep track of tasks without worrying about forgetting them!

It's
* text-based
* efficient
* ~~cheap~~**FREE**

Just follow these 3 easy steps:
1. download it from [here](https://github.com/zupey/ip/releases)
2. run it.
3. win!

Features:
- [x] Manage tasks
- [ ] Remember deadlines
- [ ] Track events
- [ ] Seamless experience

The `faster` you type, the `better` Duke is. 💯 

Sample commands:
```bash
todo Make Pull Request
bye
```

 ## Hear from our users
> This is the most life-changing application I have used - John Doe

Commit:
Add Gradle support

zupey committed:
feat: add function to create blank line

zupey committed:
feat: add echo and bye functionality

zupey committed:
feat: add user input

zupey committed:
feat: add TaskList to store tasks

zupey committed:
style: abstract out function to wrap message within two horizontal lines

zupey committed:
feat: add CheckBox class

zupey committed:
feat: add Task class and use Task to represent a task instead of String

zupey committed:
feat: add ability to mark and unmark tasks as completed

zupey committed:
feat: abstract out getStatusIcon

zupey committed:
style: rename variable "name" to "description"

zupey committed:
feat: add subclass Deadline to Task

zupey committed:
feat: add subclass Event to Task

zupey committed:
feat: add subclass ToDo to Task

zupey committed:
feat: support parsing of commands for deadline, todo and event

zupey committed:
add automated testing

zupey committed:
feat: add error handling + refactor files into packages

zupey committed:
feat: add delete task functionality

zupey committed:
feat: add enums and refactor code

refactor code into packages to abstract out handler functionality

zupey committed:
feat: add datetime support for deadline and event tasks

use LocalDateTime to specify how the date-time format the commands should be in. throws exception when user does not follow format

zupey committed:
feat: add storage functionality to save tasks between different sessions

zupey committed:
Merge branch 'add-gradle-support'

Add gradle support to the project

zupey committed:
Add Gradle support for automated tests and execution of duke

zupey committed:
Organise classes into one package `duke`

zupey committed:
Use more OOP

Abstract out the UI class which handles the UI for printing messages from duke

zupey committed:
Use more OOP

Abstract out the Parser class which handles the parsing of user input to return the arguments that are necessary for the handlers.

zupey committed:
Add checkstyle automation in gradle

Automated checkstyle of code so that it follows the module coding standard

zupey committed:
Add tests for deadline and event classes

zupey committed:
style: setup checkstyle in gradle and changed style to meet java coding standard

zupey committed:
style: edit classes in the commands package to meet coding standards

zupey committed:
style: edit classes in the exceptions package to meet coding standards

zupey committed:
fix: fixed test cases to work after shifting of packages

zupey committed:
style: change the task package to follow the java coding standard

zupey committed:
style: added javadocs for explanation of what each function does

zupey committed:
style: added javadocs for Duke.txt

zupey committed:
feat: add find functionality to find tasks

zupey committed:
merge A-Javadoc into master

resolve merge conflicts using intellij merge conflict gui

zupey committed:
Merge branch 'Level-9' into master

shaniceng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+
+        File storageDirectory = new File("./data");
+        if (!storageDirectory.exists()) {
+            if (!storageDirectory.mkdir()) {
```
Perhaps you could use this method to make your if else statement not nested for better readability :)
```suggestion
            if (!storageDirectory.mkdir() && !storageDirectory.mkdir()) {
```

zupey replied:
Thank you for your suggestion!

shaniceng commented on src/main/java/duke/commands/CommandEventHandler.java:
```diff
@@ -0,0 +1,46 @@
+     * @throws DukeException error message.
+     */
+    public void checkValid() throws DukeException {
+        if (value == null || !flag.equals("at") || additionalValue == null) {
```
Perhaps you could try to avoid complicated expressions here by initialising them first and use them in the if else statement
```suggestion
        boolean isValid = value == null || !flag.equals("at") || additionalValue == null;
        if (isValid) {
```

zupey replied:
Thanks for the suggestion!

shaniceng commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,151 @@
+        this.db = db;
+    }
+
+    public String getRemainingTasks() {
```
Great job in naming all the methods in verbs and camel cases 👍

shaniceng commented on src/main/java/duke/commands/Parser.java:
```diff
@@ -0,0 +1,40 @@
+        String value = matcher.group(2);
+        String flag = matcher.group(3);
+        String additionalValue = matcher.group(4);
+        List<String> result = new ArrayList<>();
```
Should this be "results" instead since it is a list and the name of variable should be plural
```suggestion
        List<String> results = new ArrayList<>();
```

zupey replied:
Thank you for pointing this out!

zupey committed:
Set-up JavaFX in build.gradle

zupey committed:
style: meet coding standards

add javadocs for files that were missed and change variable names to meet coding standards.

shaniceng reviewed (commented):
Overall, all of your codes are easy to read and it LGTM 👍 
Just some nits to be fixed :)
Good Job!

ngquyduc commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+        Scanner sc = new Scanner(System.in);
+        CommandHandlerFactory commandHandlerFactory = new CommandHandlerFactory();
+        String command;
+        while (!(command = sc.nextLine()).equals("bye")) {
```
Might it be a better idea to let the Parse do all the parsing commands work?

ngquyduc commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+ */
+public class Duke {
+
+    private void initialise() {
```
Perhaps abstracting this method in a class _Storage_ might be a good idea.

ngquyduc commented on src/main/java/duke/commands/CommandDeadlineHandler.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    public void checkValid() throws DukeException {
+        if (value == null || !flag.equals("by") || additionalValue == null) {
+            throw new DukeException("Correct usage: deadline return book /by 24/04/2019 1600");
```
I like the way you throw an exception and help the user to type in the correct command at the same time.

ngquyduc commented on src/test/java/duke/tasks/DeadlineTest.java:
```diff
@@ -0,0 +1,19 @@
+
+public class DeadlineTest {
+    @Test
+    public void dbRepresentationFromUserInput() {
```
```suggestion
    public void getDbRepresentationFromUserInput() {
```

ngquyduc reviewed (commented):
Overall, clean and easy-to-follow code 👍 .
However, you can consider some of my suggestions.

zupey committed:
feat: add GUI for duke

Add basic GUI that echos the user input.

zupey committed:
feat: make user commands responsive

Duke now responds to user input.

zupey committed:
feat: closes GUI when user inputs `bye`

Duke now closes when user inputs bye.

zupey committed:
add files that were undetected by intellij

zupey committed:
feat: removes old code that is unused after GUI is implemented.

Remove smiley face from error message because there is no support for it in the interface.

zupey committed:
Refactor files that are related to GUI to the gui package.

By abstracting out all the related code to the gui package, the code becomes more organised.  All the javafx-related files can be found in the gui package.

zupey committed:
Abstract out the bye command into CommandByeHandler.java.

Previously, the bye command is checked at the user input level.
This is not ideal because the implementation is not extensible.

By capitalizing on the previously built infrastructure of CommandHandler,
I abstracted out the bye command into its own class to have better
extensibility and overall cleanliness of the code.

zupey committed:
Fix bug where index out of bounds exception is not caught correctly.

When user tries to mark or delete a task that is out of bounds,
the exception is not caught correctly.
Added junit tests that specifically test this so that we do not encounter such issues in the future.

zupey committed:
Refactor based on comments given by pull request review and java coding style requirements.

zupey committed:
Refactor javadoc header comments to follow java code style guide.

zupey committed:
Add InvalidSyntaxException to better describe a subset of errors.

Change error message for InvalidSyntaxException to be more descriptive.
Update all errors that are syntax related to throw this error instead.

zupey committed:
Add assertions.

Assert validity of commands that should be true at that point.
The first line of checking should be in the Parser class.

zupey committed:
Refactor code to improve readability.

Abstract out and name variables based on their function.

zupey committed:
TaskList: Add automatic sorting to the taskList based on the date of task.

Since Todo tasks have no date, they have arbitrary max time and hence will be put at the end of the TaskList.

zupey committed:
Refactor placement of where exceptions are caught to be more readable.

zupey committed:
Add more assertions.

zupey committed:
Merge pull request #5 from zupey/branch-A-Assertions

Branch a assertions

zupey committed:
Improve code quality.

zupey committed:
Merge pull request #6 from zupey/branch-A-CodeQuality

Improve code quality.

zupey committed:
Minor UI fix

Remove blank line when displaying taskList.
Add spacing in error message for better readability.

zupey committed:
Squash bugs

Program crashes due to null exception because logic in isValid is incorrect.
Fixed to avoid this issue.

zupey committed:
Update user guide in README.md

zupey committed:
Merge pull request #8 from zupey/branch-A-UserGuide

Update user guide in README.md

zupey committed:
Merge branch 'master' of github.com:zupey/ip

zupey committed:
Update build.gradle for release of v1.0.

zupey committed:
Add UI.png :sparkles:

zupey committed:
Rename UI.png :sparkles:

zupey committed:
Minor UI fixes

Dialog box now fits the size of text sent.
Error message changed.
